File: Platforms/Common/Reign_Video_D3D11_Component/States/SamplerState.cpp
#include "pch.h"
#include "SamplerState.h"

namespace Reign_Video_D3D11_Component
{
	#pragma region Constructors
	SamplerStateError SamplerStateCom::Init(VideoCom^ video, SamplerStateDescCom^ desc)
	{
		null();
		this->video = video;

		ID3D11SamplerState* stateTEMP;
		if (FAILED(video->device->CreateSamplerState(desc->desc, &stateTEMP)))
		{
			return SamplerStateError::SampleState;
		}
		state = stateTEMP;

		return SamplerStateError::None;
	}

	SamplerStateCom::~SamplerStateCom()
	{
		if (state) state->Release();
		null();
	}

	void SamplerStateCom::null()
	{
		state = 0;
	}
	#pragma endregion

	#pragma region Methods
	void SamplerStateCom::Enable(int index)
	{
		ID3D11SamplerState* stateTEMP = state;
		video->deviceContext->PSSetSamplers(index, 1, &stateTEMP);
	}
	#pragma endregion
}

File: Platforms/Common/Reign.Video.D3D11/States/SamplerState.cs
﻿using System;
using Reign.Core;
using Reign_Video_D3D11_Component;

namespace Reign.Video.D3D11
{
	public class SamplerState : DisposableResource, ISamplerState
	{
		#region Properties
		private SamplerStateCom com;
		#endregion

		#region Constructors
		public SamplerState(IDisposableResource parent, ISamplerStateDesc desc)
		: base(parent)
		{
			try
			{
				var video = parent.FindParentOrSelfWithException<Video>();

				com = new SamplerStateCom();
				var error = com.Init(video.com, ((SamplerStateDesc)desc).com);

				if (error == SamplerStateError.SampleState) Debug.ThrowError("SampleState", "Failed to create SampleState");
			}
			catch (Exception e)
			{
				Dispose();
				throw e;
			}
		}

		public override void Dispose()
		{
			disposeChilderen();
			if (com != null)
			{
				com.Dispose();
				com = null;
			}
			base.Dispose();
		}
		#endregion

		#region Methods
		public void Enable(int index)
		{
			com.Enable(index);
		}
		#endregion
	}
}